### DISCUSSION TITLE
How to capture mouse events in a canvas

### DISCUSSION THREAD

**Discussion by HowardPWeiss** [unoplatform/uno#12521] [Q&A]
I am working on a WPF MVVM application using UnoPlatform for Linux on Ubuntu 20.04 using Visual Studio code and c#

I have a canvas on which I have drawn many lines and text blocks.  I would like to receive events when the mouse is clicked or moved anywhere on the canvas. I have tried the following

in Mainpage.xaml

<ItemsControl ItemsSource="{Binding ObjectCollection}" ItemTemplateSelector="{StaticResource DTS}">
            <ItemsControl.ItemsPanel>
                <ItemsPanelTemplate>
                    <Canvas x:Name="Gameboard" Margin="13,105"  Width="800" Height="922" Background="Red"  
                        VerticalAlignment="Top" HorizontalAlignment="Left"
                        MouseMove = "{Binding Canvas_PointerMoved}" 
                        >
                    </Canvas>
                </ItemsPanelTemplate>
            </ItemsControl.ItemsPanel>

where ObjectCollection is an ObservableCollection of Line and TextBlock objects

MainViewModel.cs

    public void Canvas_PointerMoved(object sender, PointerRoutedEventArgs e)
        {
            Debug.WriteLine("Canvas_PointerMoved");
        }

The PointerMoved method in the  MainViewModel is never invoked.

What is the correct way to implement this functionality?

**Comment by ToddGlodek** [comment]
Where do you register the event handlers, in the Mainpage _ctor, or in the DataContextChanged handler?    Does it work when you run it using WPF on Windows, but it doesn't run on Linux ??

**Reply by dr1rrb** [reply]
The event handlers can de defined only the "code behind" of your page (i.e. the `MainPage.xaml.cs` file). If you want to handle them in your `ViewModel` you will have to explicitly route them:

In your `MainPage.xaml` :
```xml
<Canvas PointerMoved="Canvas_PointerMoved" />
```

In your `MainPage.xaml.cs` (the "code behind")
```csharp
public void Canvas_PointerMoved(object sender, PointerRoutedEventArgs e)
{
     if (DataContext is MainViewModel vm)
    {
        vm.Canvas_PointerMoved(sender, e);
    }
}
```

then in your `MainViewModel.cs` you can keep your method
```csharp
public void Canvas_PointerMoved(object sender, PointerRoutedEventArgs e)
{
    Debug.WriteLine("Canvas_PointerMoved");
}
```

BTW, when using the MVVM pattern you should not forward those kind of events directly in your "ViewModel". Your VM should expose some  `ICommand` that are invoked by the view.

**Reply by ToddGlodek** [reply]
@dr1rrb is 100% about "Your VM should expose some ICommand".   Especially because your project is WPF.   If you are doing cross platform, you are eventually going to want to switch to WinUI.  You will want your VM to be reusable across every different platform type, and not contain references to WPF specific object types (Line and TextBlock objects) that are different types than WinUI.

For example WPF uses [System.Windows.Controls](https://learn.microsoft.com/en-us/dotnet/api/system.windows.controls?view=windowsdesktop-7.0).TextBlock.   Whereas, WinUI uses [Microsoft.UI.Xaml.Controls](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.controls?view=windows-app-sdk-1.3).TextBlock.    Although they are both called "TextBlock" - they are two totally distinct & inoperable classes.

**Comment by HowardPWeiss** [comment]
I agree with @dr1rrb that my "VM should expose some ICommand that is invoked by the view.".  I can implement the ICommand as a RelayCommand.  How do I route the event to my ViewModel method? I modified the example below and tried just binding PointerMoved ala

      PointerMoved = "{BInding ViewModel_Canvas_PointerMoved}"  

This results in a compile failure "error CS1001: Identifier expected"

I  expect that I need to remove <PointerMoved = "Canvas_PointerMoved"> from with the Canvas declaration and add an Interaction.Trigger something like

   <i:Interaction.Triggers>
      <i:EventTrigger EventName="PointerMoved">
        <i:InvokeCommandAction Command="{Binding ViewModel_Canvas_PointerMoved}" />
      </i:EventTrigger>
    </i:Interaction.Triggers>

but how do i define xmlns:i? If I define i: as xmlns:i="clr-namespace:System.Windows.Interactivity;assembly=System.Windows.Interactivity", the build generates "The property Object.Triggers is unknown"

Looking on the Internet, I discovered that System.Windows.Interactivity is not available any more (see https://stackoverflow.com/questions/46556240/wpfmvvm-library-system-windows-interactivity-in-not-available-anymore).  The same post initially recommends using NugetPackage System.Windows.Interactivity.WPF and then recommends using [Microsoft.Xaml.Behaviors.Wpf](https://www.nuget.org/packages/Microsoft.Xaml.Behaviors.Wpf/) because System.Windows.Interactivity.WPF is obsolete.  Finally I discovered a NuGet package Uno.Microsoft.Xaml.Behaviors.WinUI.Managed (which seems to refer only to behaviors).  

What packages do I need/want? Do I need use on Interactivity.Trigger? How do I bind to ViewModel_Canvas_PointerMoved?

Below is my current implementation which uses Event handlers in code-behind.  I have implemented ICommand for ViewModel_Canvas_PointerMoved but it is unused. I have not yet worried about how to the event Arguments to the the ViewModel event handler.
-------------------------------

MainPage.xaml

<Page
    x:Class="PointerTest.MainPage"
    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:local="using:PointerTest"
    xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
    xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
    
    mc:Ignorable="d"    
    Background="{ThemeResource ApplicationPageBackgroundThemeBrush}">

    <Page.DataContext>
        <local:MainViewModel/>
    </Page.DataContext>
       
    
    <Canvas x:Name="Gameboard" Width="800" Height="922" Background="Red"  
                        VerticalAlignment="Top" HorizontalAlignment="Left"
                        PointerCanceled = "Canvas_PointerCanceled"
                        PointerCaptureLost = "Canvas_PointerCaptureLost"
                        PointerEntered = "Canvas_PointerEntered"
                        PointerExited = "Canvas_PointerExited"
                        PointerMoved = "Canvas_PointerMoved"
                        PointerPressed = "Canvas_PointerPressed"
                        PointerReleased = "Canvas_PointerReleased"
                        PointerWheelChanged = "Canvas_PointerWheelChanged"
    >
    </Canvas>
</Page>

MainPage.xaml.cs
------------------------

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Runtime.InteropServices.WindowsRuntime;
using Windows.Foundation;
using Windows.Foundation.Collections;
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Controls;
using Microsoft.UI.Xaml.Controls.Primitives;
using Microsoft.UI.Xaml.Data;
using Microsoft.UI.Xaml.Input;
using Microsoft.UI.Xaml.Media;
using Microsoft.UI.Xaml.Navigation;
using System.Diagnostics;
using Windows.Devices.Input;

// The Blank Page item template is documented at http://go.microsoft.com/fwlink/?LinkId=402352&clcid=0x409

namespace PointerTest
{
    /// <summary>
    /// An empty page that can be used on its own or navigated to within a Frame.
    /// </summary>
    public sealed partial class MainPage : Page
    {
        public MainPage()
        {
            Debug.WriteLine("MainPage");
            this.InitializeComponent();
        }

        private void ExplainPointer(string eventName, object sender, PointerRoutedEventArgs e)
        {
            Debug.WriteLine(eventName);
            Microsoft.UI.Xaml.Input.Pointer ptr = e.Pointer;

            // Multiple, simultaneous mouse button inputs are processed here.
            // Mouse input is associated with a single pointer assigned when 
            // mouse input is first detected. 
            // Clicking additional mouse buttons (left, wheel, or right) during 
            // the interaction creates secondary associations between those buttons 
            // and the pointer through the pointer pressed event. 
            // The pointer released event is fired only when the last mouse button 
            // associated with the interaction (not necessarily the initial button) 
            // is released. 
            // Because of this exclusive association, other mouse button clicks are 
            // routed through the pointer move event.

            // Type dt = ptr.PointerDeviceType.GetType();
            // string dtTypeName = dt.FullName;

            // Debug.WriteLine(String.Format("Type of ptr.PointerDeviceType {0}", dtTypeName));

            if (ptr.PointerDeviceType == Microsoft.UI.Input.PointerDeviceType.Mouse)            
            {
               
                // To get mouse state, we need extended pointer details.
                // We get the pointer info through the getCurrentPoint method
                // of the event argument. 

                Microsoft.UI.Input.PointerPoint ptrPt = e.GetCurrentPoint(sender as Microsoft.UI.Xaml.UIElement);
               
                if (ptrPt.Properties.IsLeftButtonPressed)
                {
                    Debug.Write("Left button ");
                }
                if (ptrPt.Properties.IsMiddleButtonPressed)
                {
                    Debug.Write("Wheel button ");
                }   
                  if (ptrPt.Properties.IsRightButtonPressed)
                {
                    Debug.Write("Right button ");
                }

		        Debug.Write(String.Format("[{0}, {1}] raw [{2}, {3}] ",
                    ptrPt.Position.X, ptrPt.Position.Y, ptrPt.RawPosition.X, ptrPt.RawPosition.Y));
                Debug.WriteLine(ptrPt.PointerId);
            }
            
            // Prevent most handlers along the event route from handling the same event again.
    	    e.Handled = true;
        }

    
        private void Canvas_PointerMoved(object sender, PointerRoutedEventArgs e)
        {
            ExplainPointer("Canvas_PointerMoved", sender, e);
        }

        private void Canvas_PointerCanceled(object sender, PointerRoutedEventArgs e)
        {
            ExplainPointer("Canvas_PointerCanceled", sender, e);
        }
        
        private void Canvas_PointerCaptureLost(object sender, PointerRoutedEventArgs e)
        {
            ExplainPointer("Canvas_PointerCaptureLost", sender, e);
        }

	    private void Canvas_PointerEntered(object sender, PointerRoutedEventArgs e)
        {
            ExplainPointer("Canvas_PointerEntered", sender, e);
        }
        
        private void Canvas_PointerExited(object sender, PointerRoutedEventArgs e)
        {
            ExplainPointer("Canvas_PointerExited", sender, e);
        }
        
        private void Canvas_PointerPressed(object sender, PointerRoutedEventArgs e)
        {
            ExplainPointer("Canvas_PointerPressed", sender, e);
        }
        
        private void Canvas_PointerReleased(object sender, PointerRoutedEventArgs e)
        {
            ExplainPointer("Canvas_PointerReleased", sender, e);
        }
        
        private void Canvas_PointerWheelChanged(object sender, PointerRoutedEventArgs e)
        {
            ExplainPointer("Canvas_PointerWheelChanged", sender, e);
        }
    }
}

MainViewModel.cs
-------------------------

using System;
using System.Windows.Input;
using System.Diagnostics;
using System.Collections.ObjectModel;
using System.Collections.Generic;
using System.ComponentModel;
using System.Text;
using System.IO;
using System.Threading.Tasks;
using System.Threading;
using Windows.Storage.Pickers;
using Windows.Storage;
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Controls;
using CommunityToolkit.Mvvm.Input;

namespace PointerTest
{

    public class MainViewModel : MainViewModelBase
    {
        public MainViewModel()
        {
            Debug.WriteLine("MainViewModel");
            ViewModel_Canvas_PointerMoved = new RelayCommand(doViewModel_Canvas_PointerMoved);
        }

        ~MainViewModel()
        {
            Debug.WriteLine("~MainViewModel");
        }
        
#region ViewModel_Canvas_PointerMoved
 
        public ICommand ViewModel_Canvas_PointerMoved{get; set;}
        

        private void doViewModel_Canvas_PointerMoved()
        {
            Debug.WriteLine("doViewModel_Canvas_PointerMoved");
        }

#endregion        
    }
}

**Reply by ToddGlodek** [reply]
Ah...  it seems like you are trying to build some kind of really powerful canvas object - something more complicated than what you can do with XAML. Have you considered creating a class that inherits from canvas.  Something like this maybe.   

```
namespace PointerTest;

public class MyPowerfulCanvas : Microsoft.UI.Xaml.Controls.Canvas { ... 

    public MyPowerfulCanvas() {        

        PointerPressed += this.OnPointerPressed;
        PointerReleased += this._PointerReleased;
        PointerMoved += this._PointerMoved;

     }

    private void OnPointerPressed(object sender, PointerRoutedEventArgs e) { 
         //  Call ICommand on ViewModel
     }   

    private void _PointerReleased(object sender, PointerRoutedEventArgs e) { 
         //  Call ICommand on ViewModel
    }

 }

```
If MyPowerfulCanvas is placed on the MainPage, then it will automatically consume the same ViewModel as MainPage (unless you override it).  So 

I see "Microsoft.UI.Xaml.Controls" - so it feels like you are actually using WinUI (not WPF) - that's good.  Does it work when you run it using it on Windows, but then it doesn't run on Linux ??

**Comment by HowardPWeiss** [comment]
Todd -

Your suggestion is a good one. 

On principle, I would like to know how to create XAML which would allow me to invoke a method in the ViewModel from my View for a Canvas event.  How do I bind the Event to the Corresponding ICommand.  Do I need to use Interaction.Triggers? What NuGet package should I use and how to I access the appropriate namespace in the appropriate package?

A different question is how to achieve my goal with MVVM. The view is a gameboard has 729 elements.  That is a big XAML file

I am creating this project in Linux using Visual Studio Code.  I have unsuccessfully tried move the solution to Windows.  That will be tomorrow's problem
 
Howard Weiss

**Reply by ToddGlodek** [reply]
OK - Your questions are not UnoPlatform specific - but are general "How does WinUI work".   You might find this video from [James Montemagno](https://www.youtube.com/@JamesMontemagno) useful: [MVVM Source Generators: Never Write MVVM Boilerplate Code Again!](https://www.youtube.com/watch?v=aCxl0z04BN8).  There are a bunch of other videos in his playlists that provide a wealth of guidance.   Also, [Gerald Versluis](https://www.youtube.com/@jfversluis) has a lot of good videos in his playlists.

**Comment by HowardPWeiss** [comment]
Todd -

I watched the video [MVVM Source Generators: Never Write MVVM Boilerplate Code Again](https://www.youtube.com/watch?v=aCxl0z04BN8)from James Montemago.  Very impressive.  I will certainly use the CommunityToolkit to generate source rather than doing it manually

The video focused on generating the code in the ViewModel

I am stiil not sure howto construct/generate the XAML to pass a Canvas Event from the View to the View Model

**Reply by ToddGlodek** [reply]
That's something you could easily look up in Bing's AI Chat....  However, make sure you don't get distracted by stuff that isn't related to WinUI. 

In order to implement commanding, a ViewModel should define one or more properties of type `ICommand`. The `ICommand` interface defines two methods and one event: `Execute`, `CanExecute`, and `CanExecuteChanged`¹. You can use an `EventTrigger` in combination with `InvokeCommandAction` from the `Windows.Interactivity` namespace to bind UI events to commands in the ViewModel². Here is an example:

```xml
<ListBox ...>
  <i:Interaction.Triggers>
    <i:EventTrigger EventName="SelectionChanged">
      <i:InvokeCommandAction Command="{Binding SelectedItemChangedCommand}"/>
    </i:EventTrigger>
  </i:Interaction.Triggers>
</ListBox>
```

Source: Conversation with Bing, 6/7/2023
(1) Simplifying Events with Commanding - Xamarin Blog. https://devblogs.microsoft.com/xamarin/simplifying-events-with-commanding/.
(2) WPF Binding UI events to commands in ViewModel - Stack Overflow. https://stackoverflow.com/questions/4897775/wpf-binding-ui-events-to-commands-in-viewmodel.
(3) wpf - Simple Event Handling in MVVM - Stack Overflow. https://stackoverflow.com/questions/1926176/simple-event-handling-in-mvvm.
(4) Turn Events into Commands with Behaviors - Xamarin Blog. https://devblogs.microsoft.com/xamarin/turn-events-into-commands-behaviors/.
(5) Commanding - .NET MAUI | Microsoft Learn. https://learn.microsoft.com/en-us/dotnet/maui/fundamentals/data-binding/commanding.

**Comment by HowardPWeiss** [comment]
Todd -

Your time and advice are greatly appreciated.  

The point of my original post is that I want to a command to my ViewModel as explained in the second article you reference (WPF Binding UI events to commands in ViewModel).  This means adding an interaction trigger and referencing Systems.Windows.Interactivity. I  quote from the article

<START QUOTE>

  <ListBox ...>
      <i:Interaction.Triggers>
          <i:EventTrigger EventName="SelectionChanged">
              <i:InvokeCommandAction Command="{Binding SelectedItemChangedCommand}"/>
          </i:EventTrigger>
      </i:Interaction.Triggers>
  </ListBox>
  You can reference System.Windows.Interactivity by going Add reference > Assemblies > Extensions.
  
  And the full i namespace is: xmlns:i="clr-namespace:System.Windows.Interactivity;assembly=System.Windows.Interactivity".

<END QUOTE>

When I did this in my Linux .Net7.0 Visual Studio Code solution, the subsequent compile fails with the relatively cryptic messages "The property Object.Triggers is unknown"

With some effort I duplicated my Linux Visual Studio Code solution on Windows using Visual Studio 2022 targeting .Net7.0.  

My xaml starts with

<Page
    x:Class="PointerTestVS2022.MainPage"
    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:local="using:PointerTestVS2022"
    xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
    xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
    mc:Ignorable="d"    
    Background="{ThemeResource ApplicationPageBackgroundThemeBrush}">

If I try to add 

     xmlns:i="clr-namespace:System.Windows.Interactivity;assembly=System.Windows.Interactivity"

Visual Studio rejects the change and reports the following error
 
"Assembly 'System.Windows.Interactivity' was not found. Verify that you are not missing
an assembly reference.  Also, verify that your project andall referenced assemblies have been built"

As noted in a previous comment, this is because System.Windows.Interactivity is not included in .Net7.0

It would appear that I have to convert the event into a behavior, as outlined in the fourth article you reference "Turn Events into Commands with Behaviors".  The article does provide a sample Xamarin project, which is at least a starting point

**Reply by dr1rrb** [reply]
Hey @HowardPWeiss, the `EventTrigger` is an old pattern from WPF (In UWP/WinUI, the `EventTrigger` is about triggering visual states, cf. [ms docs](https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.eventtrigger?view=winrt-22621)).  While it had some advantages, it involves reflection which is not a great idea (for performance and for assembly trimming).

Instead you have to create a custom behavior for each event you want to convert into a command, usually only few specific events. We already have an extension in [our toolkit](https://www.nuget.org/packages/uno.Toolkit.UI) for such common events (and especially the "item clicked" in a `ListView` like in your example), cf. https://platform.uno/docs/articles/external/uno.toolkit.ui/doc/helpers/control-extensions.html

```xml
<ListView
	ItemsSource="123"
	IsItemClickEnabled="True"
	utu:CommandExtensions.Command="{Binding DoSomethingCommand}" />
```

Another option would be to use the the [`x:Bind` markup extension](https://learn.microsoft.com/en-us/windows/uwp/xaml-platform/x-bind-markup-extension) and more specifically the [function binding](https://learn.microsoft.com/en-us/windows/uwp/data-binding/function-bindings). Using it you can directly "bind method" to events (so you don't even have to create an `ICommand`, and it does not involve reflection since binding is resolve at compilation). For instance with your example:

MainPage.xaml.cs
```cs
public partial class MainPage
{
	public MainViewModel VM { get; }  = new();
}
```

MainPage.xaml
```xml
<ListView 
	x:Name="MyListView"
	SelectionMode="Single"
	IsItemClickedEnabled="True"
	ItemClicked="{x:Bind VM.DoSomething(MyListView.SelectedItem)}" 
	... />
```

> Note: You should prefer to use the `ItemClicked` event over the `SelectionChanged`, since this allow user to click twice on the same item.

BTW, we also have a port of the "interactivity" namespace in the [`Uno.Microsoft.Xaml.Behaviors.Interactivity` package](https://www.nuget.org/packages/Uno.Microsoft.Xaml.Behaviors.Interactivity), however this contains only the `Behavior` helpers as the `EventTrigger` has never been ported to UWP by the MS teams.

**Reply by HowardPWeiss** [reply]
Thank you for your guidance

How would I implement the equivalent of PassEventArgsToCommand = "True"?

**Reply by dr1rrb** [reply]
With our `CommandExtensions.Command` from the toolkit, you will get the clicked item as `parameter` of the `Execute` of your command (cf. the [remarks](https://platform.uno/docs/articles/external/uno.toolkit.ui/doc/helpers/control-extensions.html#remarks) section).

With the `x:Bind`, your method just needs to have parameters that match the type of the event handler (cf. the [event-binding](https://learn.microsoft.com/en-us/windows/uwp/xaml-platform/x-bind-markup-extension#event-binding) section)

**Reply by ToddGlodek** [reply]
@nickrandolph -   I looked at the documentation referenced by @dr1rrb (https://platform.uno/docs/articles/external/uno.toolkit.ui/doc/helpers/control-extensions.html).   I noticed that the namespace 'utu' is used in the example XAML, but isn't defined anywhere in the example. 

`xmlns:utu="using:Uno.Toolkit.UI"`

Without this makes it difficult for first time user to know how to use the sample code

**Reply by jeromelaban** [reply]
Thanks for the feedback, this should do: https://github.com/unoplatform/uno.toolkit.ui/pull/641
